docs/datatypes/tref.md

```diff
@@ -0,0 +1,176 @@
+---
+id: datatypes_tref
+title:  "TRef"
+---
+
+A `TRef[A]` is part of the `STM` bundle that represents a transactional reference that can perform reads and writes inside a transactional effect.
+
+`TRef` gives the basic machinery for others `SMT` transactional data types.
+
+`TRef[A]` where `A` is an immutable value.
+
+## Create a TRef
+
+Creating a `TRef` without committing its value:
+
+```scala mdoc:silent
+import zio.stm.{STM, TRef}
+
+val tVar: STM[Nothing, TRef[Int]] = TRef.make(10)
+```
+
+Or creating a `TRef` and immediately committing its value to extract out the value:
+
+```scala mdoc:silent
+import zio.UIO
+import zio.stm.TRef
+
+val tVar: UIO[TRef[Int]] = TRef.makeCommit(10)
+```
+
+## Retrieve the value out of a TRef
+
+By retrieving the value in a single transactional statement: 
+
+```scala mdoc:silent
+import zio.UIO
+import zio.stm.TRef
+
+val tVal: UIO[Int] = (for {
+  tRef <- TRef.make(10)
+  value <- tRef.get
+} yield value).commit
+```
+
+Or on multiple transactional statements:
+
+```scala mdoc:silent
+import zio.UIO
+import zio.stm.TRef
+
+val tVal: UIO[Int] = for {
+  tRef <- TRef.makeCommit(10)
+  value <- tRef.get.commit
+} yield value
+```
+
+## Set a value to a TRef
+
+By setting a value, it will commit a new value without any relation with the actual value at certain point in time.
+
+Setting a value in a single transactional statement:
+
+```scala mdoc:silent
+import zio.UIO
+import zio.stm.TRef
+
+val tVal: UIO[Int] = (for {
+  tRef <- TRef.make(10)
+  _ <- tRef.set(20)
+  nValue <- tRef.get
+} yield nValue).commit
+```
+
+Or on multiple transactional statement:
+
+```scala mdoc:silent
+import zio.UIO
+import zio.stm.TRef
+
+val tVal: UIO[Int] = for {
+  tRef <- TRef.makeCommit(10)
+  _ <- tRef.set(20).commit
+  value <- tRef.get.commit
+} yield value
+```
+
+## Update the value of the TRef
+
+The update function `A => A`, differently from `set`, it cares about the actual value in the `TRef` by performing changes over the value and returning the same type at certain point in time.
```

Comment:
I'd rephrase to `Update modifies the content of a reference by applying function to it`.

Reply:
Perfect, let me change to your version.